torchvision/ops/diou_loss.py

```diff
@@ -0,0 +1,51 @@
+import torch
+
+from .boxes import distance_box_iou
+
+
+def distance_box_iou_loss(
+    boxes1: torch.Tensor,
+    boxes2: torch.Tensor,
+    reduction: str = "none",
+    eps: float = 1e-7,
+) -> torch.Tensor:
+    """
+    Gradient-friendly IoU loss with an additional penalty that is non-zero when the
+    distance between boxes' centers isn't zero. Indeed, for two exactly overlapping
+    boxes, the distance IoU is the same as the IoU loss.
+    This loss is symmetric, so the boxes1 and boxes2 arguments are interchangeable.
+
+    Both sets of boxes are expected to be in ``(x1, y1, x2, y2)`` format with
+    ``0 <= x1 < x2`` and ``0 <= y1 < y2``, and The two boxes should have the
+    same dimensions.
+
+    Args:
+        boxes1 (Tensor[N, 4] or Tensor[4]): first set of boxes
+        boxes2 (Tensor[N, 4] or Tensor[4]): second set of boxes
+        reduction (string, optional): Specifies the reduction to apply to the output:
+            ``'none'`` | ``'mean'`` | ``'sum'``. ``'none'``: No reduction will be
+            applied to the output. ``'mean'``: The output will be averaged.
+            ``'sum'``: The output will be summed. Default: ``'none'``
+        eps (float, optional): small number to prevent division by zero. Default: 1e-7
+
+    Returns:
+        Tensor[]: Loss tensor with the reduction option applied.
+
+    Reference:
+        Zhaohui Zheng et. al: Distance Intersection over Union Loss:
+        https://arxiv.org/abs/1911.08287
+    """
+    if boxes1.dim() == 1 and boxes2.dim() == 1:
+        batch_boxes1 = boxes1.unsqueeze(0)
+        batch_boxes2 = boxes2.unsqueeze(0)
+        diou = distance_box_iou(batch_boxes1, batch_boxes2, eps)[0, 0]
+    else:
+        diou = distance_box_iou(boxes1, boxes2, eps)[0]
```

Comment:
@yassineAlouini I'm not sure this approach is equivalent to what we had earlier. Please correct me if I'm wrong but I understand that `distance_box_iou()` does NxM pairwise comparisons while here we just want the Nx1 comparisons. This can be given by using the diagonal but this should be extremely slow as we would be throwing away the majority of operations. 

What I had in mind is try to refactor the code at `ops.boxes` so that we can share some of the estimations. Thoughts?

cc @abhi-glitchhg because you follow a similar approach on the other PR.

Reply:
BTW if you both prefer to revert to your earlier versions of the code (which didn't reuse ops.boxes) and tackle this on separate future PRs, I'm happy to go down that route. The PRs for cIoU and dIoU has been dragging for a while and I appreciate that this can become frustrating at one point. Let me know your preference so that we make this a more enjoyable experience for both of you. Thanks a bunch for your work so far. :)

Reply:
W@datumbox , I agree with your concerns. This is not computationally efficient.

Reply:
> BTW if you both prefer to revert to your earlier versions of the code (which didn't reuse ops.boxes) and tackle this on separate future PRs,

Yeah, sounds good. :)

Reply:
> @yassineAlouini I'm not sure this approach is equivalent to what we had earlier. Please correct me if I'm wrong but I understand that `distance_box_iou()` does NxM pairwise comparisons while here we just want the Nx1 comparisons. This can be given by using the diagonal but this should be extremely slow as we would be throwing away the majority of operations.
> 
> What I had in mind is try to refactor the code at `ops.boxes` so that we can share some of the estimations. Thoughts?
> 
> cc @abhi-glitchhg because you follow a similar approach on the other PR.

That's a very good point and I might have introduced a bug by going quickly on my refactor, so thanks for pointing this out.

I can revert to previous code or keep working on this here (in this PR), both work for me. :ok_hand:

Reply:
Thanks for the flexibility! Let's revert and use the previously vetted code on the loss. We can investigate refactoring all losses to share code with `ops.boxes` on separate PRs as you originally suggested.

Reply:
@yassineAlouini From what I see you've reverted the code that reused estimations from `ops.boxes` and now use the a modified version of the loss at [Detectron2](https://github.com/facebookresearch/detectron2/blob/dfe8d368c8b7cc2be42c5c3faf9bdcc3c08257b1/detectron2/layers/losses.py#L5). BTW we should add a reference in the source code similar to [this](https://github.com/pytorch/vision/blob/main/torchvision/ops/focal_loss.py#L34), to indicate the original source.

Concerning the casting question, note that in `ops.boxes` the [_upcast method](https://github.com/pytorch/vision/blob/main/torchvision/ops/boxes.py#L218-L223) is primarily there to handle overflows of low precision values that might overflow. The main concern is integers (because those methods need to support them). In the case of losses, the [_upcast method](https://github.com/pytorch/vision/blob/main/torchvision/ops/giou_loss.py#L7-L11) works a bit different as it converts everything to floats (we don't handle integers in losses). This is crucial because some of the area estimation operations will overflow in `torch.float16`.

```
>>> torch.finfo(torch.float16)
finfo(resolution=0.001, min=-65504, max=65504, eps=0.000976562, tiny=6.10352e-05, dtype=float16)
```

So I think the safe thing to do here is to follow the same approach as in gIoU and upcast. I think we should move the method `_upcast` to `_utils` so it can be shared. This is going to be useful also for the cIoU PR that @abhi-glitchhg is working on.

cc @fmassa for visibility in case I stated something incorrect here.